Improve resolving of scrollbar-gutter style

lukewarlow opened the pull request:
#### 76362384eea16a54d1b77d63215830abef329124
<pre>
Improve resolving of scrollbar-gutter style
<a href="https://bugs.webkit.org/show_bug.cgi?id=259014">https://bugs.webkit.org/show_bug.cgi?id=259014</a>

Reviewed by Darin Adler.

Move the resolution of scrollbar-gutter styles to its own function and implement for RenderLayerScrollableArea and LocalFrameView.
This will allow implementation of scrollbar-gutter on viewports.

This patch has no behaviour changes.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollbarGutterStyle const):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::scrollbarGutterStyle const):
* Source/WebCore/platform/ScrollableArea.h:
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::verticalScrollbarWidth const):
(WebCore::RenderLayerScrollableArea::horizontalScrollbarHeight const):
(WebCore::RenderLayerScrollableArea::scrollbarGutterStyle const):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:

Canonical link: <a href="https://commits.webkit.org/265882@main">https://commits.webkit.org/265882@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0cd1553827c2e32a34ea0ba13f75f6b7f0d32102

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12140 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12487 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12789 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13886 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11715 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12171 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14902 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12501 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14402 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12304 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13112 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10281 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14302 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10398 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11022 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18139 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11479 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11181 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14357 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11670 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9621 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10886 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2978 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15212 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11523 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
```diff
@@ -1023,6 +1019,13 @@ OverscrollBehavior RenderLayerScrollableArea::verticalOverscrollBehavior() const
 
+ScrollbarGutter RenderLayerScrollableArea::scrollbarGutterStyle()  const
+{
+    if (m_layer.renderBox())
```
I think it would be better style to put this into a local and use it. Can still do that inside the `if` condition.

lukewarlow replied:
Apologies I'm not sure what you mean?

darinadler replied:
```
if (auto* renderer = m_layer->renderBox())
    return renderer->style() etc.
```

lukewarlow replied:
Ah yes that makes sense apologies I was confused because it's not actually renderBox that is called inside of the if statement but there's no reason it can't be. Begs the question why the overscrollBehaviour properties are written the way they are? (scrollbarWidthStyle is the way it is because I copied those)

darinadler commented on Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
```diff
@@ -998,10 +996,8 @@ int RenderLayerScrollableArea::horizontalScrollbarHeight(OverlayScrollbarSizeRel
 
-    if (!m_hBar && m_layer.renderBox() && !(m_layer.renderer().style().scrollbarGutter().isAuto
-        || ScrollbarTheme::theme().usesOverlayScrollbars()) && !isHorizontalWritingMode) {
+    if (!m_hBar && !(scrollbarGutterStyle().isAuto || ScrollbarTheme::theme().usesOverlayScrollbars()) && !isHorizontalWritingMode)
```
It’s slightly better for performance to put tests likely to fail first when it’s a series separated by `&&` and the computations are not trivial. I am thinking `!isHorizontalWritingMode` should be earlier to save at least a virtual function call.

lukewarlow replied:
Will make that change now

darinadler commented on Source/WebCore/platform/ScrollableArea.h:
```diff
@@ -31,6 +31,7 @@
 #include "ScrollSnapOffsetsInfo.h"
 #include "ScrollTypes.h"
 #include "Scrollbar.h"
+#include "ScrollbarGutter.h"
```
It’s unfortunate we have to add a header include. Lots of files include `ScollableArea.h` and this will slow down the build. If the default implementation of the virtual function was not inlined in the header then we could just use a forward declaration here and include the header only where it’s used.

lukewarlow replied:
Will make that change now

darinadler commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -6307,6 +6307,15 @@ OverscrollBehavior LocalFrameView::verticalOverscrollBehavior()  const
+{
+    auto* document = m_frame->document();
+    auto scrollingObject = document && document->documentElement() ? document->documentElement()->renderer() : nullptr;
+    if (scrollingObject && renderView())
```
The `renderView` check is surprising enough that it needs a comment I think.

lukewarlow replied:
I'll be honest I don't actually know what that checks for. This code is a copy of scrollbarWidthStyle which was a copy of verticalOverscrollBehavior.

lukewarlow replied:
Will compile locally without it for both gutter and width and see if it breaks anything

lukewarlow replied:
Doesn't seem to be important will remove it. Will also make a follow up to see if its needed for scrollbar width (just tested and it doesn't seem to be) or the overscrollBehavior functions

lukewarlow replied:
https://bugs.webkit.org/show_bug.cgi?id=259018

lukewarlow commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
```diff
@@ -2500,6 +2501,11 @@ static bool getEventTypeFromWebEvent(const WebEvent& event, NSEventType& eventTy
+ScrollbarGutter PDFPlugin::scrollbarGutterStyle() const
+{
+    return { };
+}
```
I would have thought this could just use the parent implementation but this was necessary to avoid a build issue

```
Undefined symbols for architecture arm64:
  "__ZNK7WebCore14ScrollableArea20scrollbarGutterStyleEv", referenced from:
      __ZTVN6WebKit9PDFPluginE in UnifiedSource59-mm.o
ld: symbol(s) not found for architecture arm64
```

darinadler replied:
That can be fixed by adding WEBCORE_EXPORT to the function declaration in the base class instead.

lukewarlow replied:
Fixed that now

lukewarlow committed:
Improve resolving of scrollbar-gutter style
https://bugs.webkit.org/show_bug.cgi?id=259014

Reviewed by Darin Adler.

Move the resolution of scrollbar-gutter styles to its own function and implement for RenderLayerScrollableArea and LocalFrameView.
This will allow implementation of scrollbar-gutter on viewports.

This patch has no behaviour changes.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollbarGutterStyle const):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::scrollbarGutterStyle const):
* Source/WebCore/platform/ScrollableArea.h:
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::verticalScrollbarWidth const):
(WebCore::RenderLayerScrollableArea::horizontalScrollbarHeight const):
(WebCore::RenderLayerScrollableArea::scrollbarGutterStyle const):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:

Canonical link: https://commits.webkit.org/265882@main
